### ISSUE TITLE
Storing view-settings (like video quality choice) so streams are started with one click

### ISSUE THREAD

**Issue by mYnDstrEAm** [rols1/Kodi-Addon-ARDundZDF#16]
Having to do a lot of clicking and waiting to start watching content in the media libraries is very inconvenient, it would be great if view-settings were stored and automatically selected when you select a content instead of e.g. having to click ->Streaming sources->one of the sources.

This setting could get changed in the settings. Things could remain as is at first but show a popup asking the user something like "Set this as default view-setting?" in a prompt that's shown once the stream has started.

One could also add functionality to detect streaming-quality issues such as buffering or unnecessarily low video-quality settings to show a prompt asking for a change to the default setting. But that's probably far too advanced for now and would be a separate issue.

Every click takes quite a while to load and this is currently the main problem I found in the addon itself.

**Comment by mYnDstrEAm** [comment]
Just found Sofort-Start / immediate start EIN/AUS with ctrl+c. Could this be turned on by default? And which settings are made default with this - I could not find any video-quality settings in the config.

**Comment by rols1** [comment]
> Just found Sofort-Start / immediate start EIN/AUS with ctrl+c. Could this be turned on by default? And which settings are made default with this - I could not find any video-quality settings in the config.

- the default general setting of "Sofort-Start" can be found in Addon Settings, category "Allgemein", setting ""TV-Live / Videos: Sofortstart statt Auflösungen und Download". 
- setting video format is done here:  Addon Settings, category "Modul-Einstellungen", setting "gesamtes Addon: Videoformat bei Sofortstart"
- setting video quality  is done here:  Addon Settings, category "Modul-Einstellungen", setting "gesamtes Addon: Videoqualität bei Sofortstart"

a special setting is available for modul funk.
More about  settings could be found 

More about settings can be found in the [addon wiki](https://github.com/rols1/Kodi-Addon-ARDundZDF/wiki)
/R

**Comment by mYnDstrEAm** [comment]
Thanks!

Could Sofort-Start be enabled by default after installing the addon?

Afterwards (maybe before this goes into a release) there could also be a way to make it select the most appropriate resolution. For example, by letting the user select it during the first launch of stream (this approach has problems such as that the choice may vary per channel or that one would need to wait for the stream to run for a while so the user can assess which resolution fits best) or by doing some testing to set it automatically or by using a setting that is most likely at least sufficient for the user (and maybe there already is a setting for video quality in Kodi set which can simply be copied) or a default setting like 720p which gets changed to AUTO in all cases where lots of buffering is detected.

**Comment by rols1** [comment]
> Could Sofort-Start be enabled by default after installing the addon?
yes  - changed in upcoming release

> One could also add functionality to detect streaming-quality issues such as buffering or unnecessarily low video-quality settings to show a prompt asking for a change to the default setting. But that's probably far too advanced for now and would be a separate issue.

> Afterwards (maybe before this goes into a release) there could also be a way to make it select the most appropriate resolution. For example, by letting the user select it during the first launch of stream (this approach has problems such as that the choice may vary per channel or that one would need to wait for the stream to run for a while so the user can assess which resolution fits best) or by doing some testing to set it automatically or by using a setting that is most likely at least sufficient for the user (and maybe there already is a setting for video quality in Kodi set which can simply be copied) or a default setting like 720p which gets changed to AUTO in all cases where lots of buffering is detected.

at this time, this can't be implemented by the addon. Of course, there are api-functions available to interact between addon and Kodi-player (seek time, pause, stop, isplaying, ..) but none of them to change the stream properties.
Maybe something similar could be done with ffprobe (part of ffmpeg), but I think that can be implemented on all different Kodi platforms (e.g. Android). Kodi itself uses ffmpeg as a compiled binary that cannot be affected by addons.
/R

**Comment by mYnDstrEAm** [comment]
Amazing, thank you!

> but none of them to change the stream properties.

But wouldn't it change the option in the addon settings rather than the stream properties? Maybe the stream needs to be restarted once the option has been changed. Maybe you referred to the detection of stream problems where its detects buffering/stuttering/pixelated video? In case where things don't work on all platforms it would still be useful to implement it when dependencies are available / for the platforms on which it works (those on which it doesn't could get this or an equivalent implemented later on).

**Comment by rols1** [comment]
pardon - I think there is a fundamental misunderstanding here. The addon first determines the streams (HLS, MP4) provided by the broadcaster for a video. These are usually available in different resolutions. Before starting the Kodi player, the addon selects the stream that best fits the settings. Finally, the addon passes the selected stream to the Kodi player. Usually, the addon does not influence the playing of the video afterwards.

So it is not that the player is asked by the addon to play the video with a certain property, but the addon must first find the desired stream and then pass its address.

With adaptive streams (HLS) there is another special feature: they contain a set of resolutions in the so-called master url. Here the player or the compiled-in ffmpeg try to find the resolution that fits the used hardware before starting. Therefore HLS streams usually take a bit longer to start.

How large the number of available streams for a video can be, can be observed well with the setting Sofortstart OFF.
In the internal playlist of the addon, the start time of the video may be reset afterwards - as far as I know at the moment, this is the only case where the player is controlled by the addon.
 /R

**Comment by rols1** [comment]
I feel a strong motivation behind your requests, but I fear that the github-issues do not achieve the desired reach. 
The most attention to topics around this addon can be found in the forum kodinerds (in German).
I will reply to issue #15 tomorrow. Since it concerns the existence of similar addons, I want to sleep on it first.
/R

**Comment by mYnDstrEAm** [comment]
>So it is not that the player is asked by the addon to play the video with a certain property, but the addon must first find the desired stream and then pass its address.

I never thought this would be the case. I tried to clarify this with "But wouldn't it change the option in the addon settings rather than the stream properties?" and with "select the most appropriate resolution" I was referring to selecting the most appropriate resolution stream. When changing the resolution because it detected that it's buffering too much (for example, this is just one of the possible ways this could be done) it would select another stream source.

>With adaptive streams (HLS) there is another special feature: they contain a set of resolutions in the so-called master url. Here the player or the compiled-in ffmpeg try to find the resolution that fits the used hardware before starting. Therefore HLS streams usually take a bit longer to start.

Ah okay, that's interesting...then in cases where HLS is available (I don't know if that applies to all public broadcasters) this could already be solved by simply setting HLS as the default option (which can of course still be changed in the options or become a fallback option if there are problems with the user-configured option).

>In the internal playlist of the addon, the start time of the video may be reset afterwards

I think you meant to say that it sets the start time again so you can continue where you left off, not that it _reset_ the start time to 0. Just to prevent confusion if some other dev reads this.

>but I fear that the github-issues do not achieve the desired reach.
The most attention to topics around this addon can be found in the forum kodinerds (in German).

I just think the format is much better. You can track individual issues rather than having everything mixed up in a long forum thread where you can't really keep track of things. Here you can close and subscribe to issues and revisit it long time afterwards with all the relevant info at once place. Not having that much of a reach is a feature if it's lengthy discussions but it's a good idea to post there once some discussion has taken place here. Maybe I should write in German as afaik the addon is _currently_ only of interest to German-speaking audience anyway. However, I don't know if there ever will be other language content in these media libraries or already are so it may also be of international interest. I'll probably post there at some point with a links to issues here with brief info about it (rather than our entire discussion) and info about what would need to be solved to add the proposed features.

**Comment by rols1** [comment]
> Not having that much of a reach is a feature if it's lengthy discussions but it's a good idea to post there once some discussion has taken place here. Maybe I should write in German as afaik the addon is _currently_ only of interest to German-speaking audience anyway. However, I don't know if there ever will be other language content in these media libraries or already are so it may also be of international interest. I'll probably post there at some point with a links to issues here with brief info about it (rather than our entire discussion) and info about what would need to be solved to add the proposed features.

Den Gedanken des Sprachwechsels nehme ich gerne auf und antworte künftig auf deutsch, so auch für [issue#17](https://github.com/rols1/Kodi-Addon-ARDundZDF/issues/17) (heute abend). Mit meinem Englisch ist es ohnehin nicht weit her und Missverständnisse lassen sich so reduzieren. Bei Bedarf ist ja jederzeit ein Wechsel möglich. Zur Zeit haben aber die internationalen Inhalte (Menü ZDFinternational, Nachrichteninhalte in ukrainisch/russisch) keine Einfluss auf die Addon-Entwicklung.
Ich hoffe, es ist OK, wenn ich für die Anrede das Du verwende - dem Umgang bei kodinerds entsprechend.



> I think you meant to say that it sets the start time again so you can continue where you left off, not that it _reset_ the start time to 0. Just to prevent confusion if some other dev reads this.

Ja, das ist richtig.

> Ah okay, that's interesting...then in cases where HLS is available (I don't know if that applies to all public broadcasters) this could already be solved by simply setting HLS as the default option (which can of course still be changed in the options or become a fallback option if there are problems with the user-configured option).

soweit ich es dem bisherigen Support entnehme, ist dies auch häufig der Fall. Das MP4-Format kommt bei diesen Nutzern dann nur für den Download in Frage ( Sofortstart auf AUS). Wieviele MP4-Streams angezeigt werden, richtet sich dabei nach dem Setting "alle verfuegbaren Video-Qualitaeten anbieten". 

Generell denke ich, dass zum Erreichen einer optimalen Streamqualität der Fokus nicht allein auf dem Setting liegen sollte. Hier gibt es ja noch andere beeinflussende Faktoren, wie LAN-Umgebung, Provider-Anbindung, RAM-/CPU-Auslastung usw.
/R

**Comment by rols1** [comment]
### Prüfung Umsetzung "Streamqualitätscheck zur Laufzeit" / Dialog "Set this as default view-setting?"

**1. Umsetzung "Streamqualitätscheck zur Laufzeit"**: addonseitig nicht machbar (alleinige Kontrolle durch Kodi-Player nach Übergabe des Streams durch das Addon)

**2. Dialog "Set this as default view-setting?"**
Der Aufwand für die Umsetzung wäre aus Entw.-Sicht nicht gerechtfertigt. Gründe:
- eine Erkennung für den Wechsel der früher verwendeten Konfig wäre erforderlich 
- eine Option "diesen Hinweis nicht mehr zeigen" wäre erforderlich und müsste im Addon implementiert werden (da von Kodi-Settings nicht unterstützt)
- die Addon-Settings enthalten bereits die adaptive Voreinstellung, also die autom. Anpassung: Setting Videoformat "HLS (adaptiv)", Setting Videoqualität "auto (nur HLS)
- bei akuten Stream-Problemen (Stocken, Aussetzen, Abbruch) sollte der Nutzer die Fehlersuche nicht auf die Settings beschränken, sondern auf vorrangig Netzwerk und Provideranbindung prüfen. Dazu können die bei den Einzelstreams im Addon angezeigten Bitraten zum Abgleich herangezogen werden. 
- alternativen Ansatz verworfen: Anbieten eines Tools zum Herausfinden einer optimalen Streamqualität (ffprobe o.ä. nicht Bestandteil von Kodi, nicht überall zusätzlich verfügbar oder nur aufwändig installier- und nutzbar)

Demgegenüber stehen auf der Nutzerseite die wenigen Setting-Klicks, mit denen die verfügbaren MP4-Formate durchprobiert werden können, falls es mit HLS/auto nicht zufriedenstellend funktioniert.
Und schließlich bleibt noch die Möglichkeit der Support-Anfrage mit einem Debug-Log. 100%iger Komfort ist mit einem Kodi-Addon leider nicht zu verwirklichen. 
/R
